src/client.js

```diff
@@ -0,0 +1,275 @@
+    // under us.  It's still possible for a properly motivated person
+    // to reset the mtime and size to what we expect, but these checks
+    // are more about non-intentional mistakes.  We cannot compare hashes before
+    // and after since we're computing the hash 
+    let finishedstats = await fs.stat(filename);
+    if (size !== filestats.size) {
+      throw new Error('File has a different number of bytes than was hashed');
+    } else if (finishedstats.size !== filestats.size) {
+      reject(new Error('File changed size during preperation'));
+    } else if (finishedstats.mtime.getTime() !== filestats.mtime.getTime()){
+      reject(new Error('File was modified during preperation'));
+    } else if (finishedstats.ino !== filestats.ino){
+      reject(new Error('File has changed inodes'));
+    } else {
+      return {filename, sha256, size, parts};
+    }
+  }
+
+  __useMulti(size, forceMP, forceSP) {
+    // We want the ability to force multi or single
+    forceMP = process.env.FORCE_MP || forceMP;
+    forceSP = process.env.FORCE_SP || forceSP;
+
+    if (!size) {
+      throw new Error('You must provide a size');
+    }
+
+    if (forceMP && forceSP) {
+      throw new Error('Forcing singlepart and multipart is mutually exclusive');
+    } else if (forceMP) {
+      return true;
+    } else if (forceSP) {
+      return false;
+    } else if (size >= this.multisize) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  async prepareUpload(opts) {
+    opts = runSchema(opts, Joi.object().keys({
+      filename: Joi.string().required(),
+      forceMP: Joi.boolean().truthy(),
+      forceSP: Joi.boolean().truthy(),
+      partsize: Joi.number().max(5 * GB).default(this.partsize),
+    }).without('forceSP', 'forceMP'));
+    let {filename, partsize, forceMP, forceSP} = opts;
+
+    let filesize = (await fs.stat(filename)).size;
+    if (typeof filesize !== 'number') {
+      throw new Error('Unable to determine filesize of ' + filename);
+    }
+
+    if (this.__useMulti(filesize, forceMP, forceSP)) {
+      return this.__prepareMultipartUpload({filename, partsize});
+    } else {
+      return this.__prepareSinglepartUpload({filename});
+    }
+  }
+
+
+  /**
+   * Take the list of requests in interchange format and 
+   * run them using the information from the upload preperation
+   * list
+   */
+  async runUpload(request, upload) {
+    upload = runSchema(upload, Joi.object().keys({
+      filename: Joi.string().required(),
+      sha256: schemas.sha256.required(),
+      size: Joi.number().required(),
+      parts: schemas.parts.required(),
+    }).optionalKeys('parts'));
+
+    let {filename, sha256, size, parts} = upload;
+    let etags = [];
+    let responses = [];
+
+    if (!Array.isArray(request)) {
+      request = [request];
+    }
+
+    for (let req of request) {
+      await InterchangeFormat.validate(req);
+    }
+
+    if (!parts) {
```

Comment:
That implies that all parts are exactly the same, which is not strictly true.  We could have parts where the first is 5mb, then 6mb then 7mb, then 5gb then back to 5mb.  Rather than having to iterate through the list to sum up all the preceeding element sizes, I'd rather be able to iteration and create the filestreams based on the object in the collection.